lib/dest/writeContents/index.js

```diff
@@ -36,27 +100,65 @@ function writeContents(writePath, file, cb) {
     cb(err, file);
   }
 
-  function written(err) {
-
-    if (isErrorFatal(err)) {
-      return complete(err);
+  function written(err, fd) {
+    function complete2(err1) {
+      fs.close(fd, function(err2) {
+        complete(err1 || err2);
+      });
     }
 
-    if (!file.stat || typeof file.stat.mode !== 'number' || file.symlink) {
-      return complete();
+    function applyStatFd(err, st) {
+      if (err) {
+        return complete2(err);
+      }
+
+      var mode = getMode(st, file.stat);
+      if(mode == null) {
+        return futimes(fd, st, file.stat, complete2);
+      }
+
+      fs.fchmod(fd, mode, function(error) {
+        if (error) {
+          return complete2(error);
+        }
+
+        futimes(fd, st, file.stat, complete2);
+      });
     }
 
-    fs.stat(writePath, function(err, st) {
+    function applyStatPath(err, st) {
       if (err) {
         return complete(err);
       }
-      var currentMode = (st.mode & parseInt('0777', 8));
-      var expectedMode = (file.stat.mode & parseInt('0777', 8));
-      if (currentMode === expectedMode) {
-        return complete();
+
+      var mode = getMode(st, file.stat);
+      if(mode == null) {
+        return utimes(writePath, st, file.stat, complete);
       }
-      fs.chmod(writePath, expectedMode, complete);
-    });
+
+      fs.chmod(writePath, mode, function(error) {
+        if (error) {
+          return complete(error);
+        }
+
+        utimes(writePath, st, file.stat, complete);
+      });
+    }
+
+
+    if (isErrorFatal(err)) {
+      return complete(err);
+    }
+
+    if (!file.stat) {
+      return complete();
+    }
+
+    if (typeof fd === 'number') {
+      return fs.fstat(fd, applyStatFd);
```

Comment:
Wow, that's surprising... Maybe the point is that you are fetching a file
descriptor for the directory and not trying to use some fancy functions
like opendir() (that's not available on Node.js yet...)? What version of
Node.js are you using? I'm afraid it would only works on the latests ones
or we are using here an implementation bug, but if there are no more
concerns, I'm +1 to use that hack if it's working for our purposses :-)
El 07/12/2015 21:46, "Blaine Bublitz" <notifications@github.com> escribió:

> In lib/dest/writeContents/index.js
> <https://github.com/gulpjs/vinyl-fs/pull/128#discussion_r46874770>:
>
> > +
> > +        utimes(writePath, st, file.stat, complete);
> > +      });
> > +    }
> > +
> > +
> > +    if (isErrorFatal(err)) {
> > +      return complete(err);
> > +    }
> > +
> > +    if (!file.stat) {
> > +      return complete();
> > +    }
> > +
> > +    if (typeof fd === 'number') {
> > +      return fs.fstat(fd, applyStatFd);
>
> @piranna <https://github.com/piranna> just tested fs.fchmod on a
> read-only file descriptor to a directory and it works. Could you simplify
> this PR by only using file descriptors?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gulpjs/vinyl-fs/pull/128/files#r46874770>.
>